### PR TITLE
Random game picker API

### PR DESCRIPTION
**Pull request by sgstair**
This is the API side for #1194

This PR introduces a new API to get a list of random games. It's been designed to not hit the database a lot (mainly caches the data in APCu) and should be acceptably fast. Details below:
* New random games GET API (~~`/vx/randomgames/get`~~ `/vx/random/game/get`) can query for up to 50 random games, returned as a node id list in `games`. There's also another returned value `listenabled` which will be true and the games list non-empty only when the event is in its Voting or Complete phases.
* Query with `?count=n` to get a specific number of games, otherwise the default is 5
* Provides a mechanism to filter, you can query for ~~`/vx/randomgames/get/compo`~~ `/vx/random/game/get/compo` or `/jam` (other values are meaningless currently)
* Data from the last simulation run shows calls take on average `1.02ms` (considerably better than most calls in the API), and max of `3.9ms` (when generating the cached structure, given the 40 or so games on the simulation event)

As for operation:
* `node_randomgames.php` will, on request, generate a structure containing a list of all games in the current event, in a minimal format. It does two custom DB queries to retrieve only as much data as it needs. It then computes a list of games and weights, games with a high enough weight go into a weighted random list, any other games go into an unweighted list which games are picked from randomly if the weighted random list runs out.
* Once the structure is generated, it's cached for 3 minutes to reduce DB load. This structure is invalidated at the end of magic.php so fresh data will immediately be computed from the updated smart ratings.
* From the generated data structure, the code uses a weighted random selection algorithm to pick games based on their weight. In the future we can boost these weights based on private information to help ensure the games with highest scores will get more attention, without revealing this information (provided that people use this mechanism to select games) - This aspect is #1193. If the system runs out of games with weights, it will pad the list with unweighted games.

Unrelated changes:
* Simulator is now adding more markdown styles into its text.

### CODE REVIEW

File: public-api/vx/randomgames.php [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+		
+		// If there is a user id, find the user's games and exclude them
+		$excludegames = [];
+		if ( $user_id = userAuth_GetID() ) {		
+			$userauthoredlinks = nodeLink_GetByKeyNode("author", $user_id);
+			foreach( $userauthoredlinks as $authorlink ) {
+				// This will only go one way, but guard anyway.
+				if ( $authorlink['b'] == $user_id ) { $excludegames[] = $authorlink['a']; }
+			}
+		}
```
Comment (local-minimum) [inline public-api/vx/randomgames.php]: This does not exclude games voted on right? That would be a nice filter to have too to increase the usefulness.

Comment (sgstair) [reply public-api/vx/randomgames.php]: It excludes games you created :) That's a good point though, and I can certainly add that.

**Comment by sgstair** [comment]
Fixed both issues that were brought up.
* `/randomgames/get` no longer includes games that you have voted on
* New API `/grade/getgames` (or `/grade/getgames/[event node ID]`) that returns `event_id` and `games` as a list of node IDs your user has voted on.

**Comment by local-minimum** [comment]
I'm not sure what the comment list is in this context, but my plan was to soonish add a games I have voted on on the featured event. But I figured it should be close to the notifications. So /home/votehistory or something

**Comment by sgstair** [comment]
re: comment list - I mean I have long planned to build a page to list comments you've posted. Not exactly the same as games you've commented on,

**Comment by Noojuno** [comment]
It might be better to make it `/vx/random/get/` with `type`, `subtype`, and `subsubtype` arguments in case it's wanted for any other type of node that isn't a game. I think the more generic an api call for this sort of thing the better.

**Comment by local-minimum** [comment]
(comments list I'd like too. `/home/history/comments` I'd rather place it cuz it isn't about seeing what someone else had commented cross the site, but to see you own.)

**Comment by sgstair** [comment]
@noojuno It's very very not generic; I can't imagine returning random other kinds of nodes, that's why I chose to name it like this.

@local-minimum I think it's useful for the comment history (minus anonymous comments) to be visible to other users also.

**Comment by sgstair** [comment]
still thinking over the rename of /randomgames, but I'm leaning towards the opinion that it's appropriate. Of course if @povrazor has an opinion I may have to reconsider :P

**Comment by Noojuno** [comment]
@sgstair I could see potential scenarios where a random user or a random video (if they're implemented in future) could be useful. Same as a random tag or similar

**Comment by povrazor** [comment]
Definitely de-game it.

**Comment by sgstair** [comment]
Renamed the endpoint to /random/game - If you mean something further please clarify. As it is the code is necessarily very game specific. If we want to do a more generic random picker in the future, I think we have some more design to do.

**Comment by povrazor** [comment]
Yeah, this is fine for now.

**Pull request merged**